src/python/pants/rules/core/binary.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from dataclasses import dataclass
+from typing import Any
+
+from pants.backend.python.rules.pex import Pex
+from pants.build_graph.address import Address
+from pants.engine.addressable import BuildFileAddresses
+from pants.engine.console import Console
+from pants.engine.fs import DirectoryToMaterialize, Workspace
+from pants.engine.goal import Goal, LineOriented
+from pants.engine.legacy.graph import HydratedTarget
+from pants.engine.legacy.structs import PythonBinaryAdaptor
+from pants.engine.rules import console_rule, optionable_rule, rule, union
+from pants.engine.selectors import Get
+
+
+@dataclass(frozen=True)
+class Binary(LineOriented, Goal):
+  name = 'binary'
+
+
+@union
+@dataclass(frozen=True)
+class BinaryTarget:
+  pass
+
+
+@union
+@dataclass(frozen=True)
+class CreatedBinary:
+  _inner: Any
```

Comment:
So, rather than `Any`, this should probably just be a `Digest`?

Can take a look at how the new `fmt` goal does this: by emitting and consuming a FmtResult: https://github.com/pantsbuild/pants/blob/6c1c3be4679eb1859be9dfdb7b01a21842c6298f/src/python/pants/rules/core/fmt.py#L23-L27